streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -103,36 +104,46 @@ public void validatePartitionExists(final TopicPartition topicPartition, final S
     public void register(final StateRestorer restorer) {
         restorer.setGlobalRestoreListener(stateRestoreListener);
         stateRestorers.put(restorer.partition(), restorer);
+        // if it has already been initialized which can happen if there was a lock exception
+        // so some tasks may have been restored
+        if (initialized) {
+            addedPostInitialization.put(restorer.partition(), restorer);
+        }
     }
 
 
     public Collection<TopicPartition> restore() {
-        final List<TopicPartition> completed = new ArrayList<>();
         if (!initialized) {
-            completed.addAll(initialize());
+            needsRestoring.clear();
```

Comment:
This is a meta comment: I found this initialization and postInitilization logic a bit messed. How about the following:

1. We keep three maps: `stateRestorers` keeps all the restorers ever registered; `needsInitilization` keeps all the restorers registered within this iteration; `needsRestoring` keeps all the initialized restorers that are not completed yet.

2. In `register()`, always add the restorer to both `stateRestorers` and `needsInitialization`.

3. Then at the start of this `restore()` function, we just check if `needsInitialization` is empty or not, if not call `initialize(needsInitialization)` and then clear `needsInitialization` map; 

4. we do not need to clear `needsRestoring` map at `restore` since it will always be cleared upon `reset`. So during rebalance the `needsRestoring` will always start as empty map and gradually grows and then shrinks; we can remove the `initialized` flag as well.

Reply:
+1

Reply:
sounds good